src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs

```diff
@@ -0,0 +1,341 @@
+                    object deserialized = await JsonSerializer.DeserializeAsync(memoryStream, type, options);
+                    AssertJsonEqual(json, JsonSerializer.Serialize(deserialized, options));
+                }
+                catch (NotSupportedException ex)
+                {
+                    Assert.True(GetTypesNotSupportedForDeserialization(elementType).Contains(type));
+                    Assert.Contains(type.ToString(), ex.ToString());
+                }
+            }
+
+            // Deserialize with extra whitespace
+            string jsonWithWhiteSpace = GetPayloadWithWhiteSpace(json);
+
+            using (var memoryStream = new MemoryStream(Encoding.UTF8.GetBytes(jsonWithWhiteSpace)))
+            {
+                try
+                {
+                    object deserialized = await JsonSerializer.DeserializeAsync(memoryStream, type, options);
+                    AssertJsonEqual(json, JsonSerializer.Serialize(deserialized, options));
+                }
+                catch (NotSupportedException ex)
+                {
+                    Assert.True(GetTypesNotSupportedForDeserialization(elementType).Contains(type));
+                    Assert.Contains(type.ToString(), ex.ToString());
+                }
+            }
+        }
+
+        private static object GetPopulatedCollection<TElement>(Type type, int stringLength)
+        {
+            TElement value = GetCollectionElement<TElement>(stringLength);
+
+            if (type == typeof(TElement[]))
+            {
+                return GetArr_TypedElements(value);
+            }
+            else if (type == typeof(ImmutableList<TElement>))
+            {
+                return ImmutableList.CreateRange(GetArr_TypedElements(value));
+            }
+            else if (type == typeof(ImmutableStack<TElement>))
+            {
+                return ImmutableStack.CreateRange(GetArr_TypedElements(value));
+            }
+            else if (type == typeof(ImmutableDictionary<string, TElement>))
+            {
+                return ImmutableDictionary.CreateRange(GetDict_TypedElements(value));
+            }
+            else if (
+                typeof(IDictionary<string, TElement>).IsAssignableFrom(type) ||
+                typeof(IReadOnlyDictionary<string, TElement>).IsAssignableFrom(type) ||
+                typeof(IDictionary).IsAssignableFrom(type))
+            {
+                return Activator.CreateInstance(type, new object[] { GetDict_TypedElements(value) });
+            }
+            else if (typeof(IEnumerable<TElement>).IsAssignableFrom(type))
+            {
+                return Activator.CreateInstance(type, new object[] { GetArr_TypedElements(value) });
+            }
+            else
+            {
+                return Activator.CreateInstance(type, new object[] { GetArr_BoxedElements(value) });
+            }
+        }
+
+        private static void AssertJsonEqual(string expected, string actual)
```

Comment:
Is this necessary due to reflection ordering being nondeterministic?

I wonder if this should be moved to a shared test utility class.

Reply:
Yes, due to nondeterministic reflection.

I've added this to `JsonTestHelper` as this is useful when verifying serialized payloads with multiple properties (nicer than bunch of `Assert.Contains` statements). Also see https://github.com/dotnet/runtime/pull/35046#discussion_r416060333.